diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/NullableTypeRule.kt

```diff
@@ -0,0 +1,140 @@
+import com.pinterest.ktlint.core.ast.ElementType.CALL_EXPRESSION
+import com.pinterest.ktlint.core.ast.ElementType.CHARACTER_CONSTANT
+import com.pinterest.ktlint.core.ast.ElementType.CHARACTER_LITERAL
+import com.pinterest.ktlint.core.ast.ElementType.CLOSING_QUOTE
+import com.pinterest.ktlint.core.ast.ElementType.DOT_QUALIFIED_EXPRESSION
+import com.pinterest.ktlint.core.ast.ElementType.EQ
+import com.pinterest.ktlint.core.ast.ElementType.FLOAT_CONSTANT
+import com.pinterest.ktlint.core.ast.ElementType.FLOAT_LITERAL
+import com.pinterest.ktlint.core.ast.ElementType.INTEGER_CONSTANT
+import com.pinterest.ktlint.core.ast.ElementType.INTEGER_LITERAL
+import com.pinterest.ktlint.core.ast.ElementType.NULL
+import com.pinterest.ktlint.core.ast.ElementType.NULLABLE_TYPE
+import com.pinterest.ktlint.core.ast.ElementType.OPEN_QUOTE
+import com.pinterest.ktlint.core.ast.ElementType.PROPERTY
+import com.pinterest.ktlint.core.ast.ElementType.QUEST
+import com.pinterest.ktlint.core.ast.ElementType.REFERENCE_EXPRESSION
+import com.pinterest.ktlint.core.ast.ElementType.STRING_TEMPLATE
+import com.pinterest.ktlint.core.ast.ElementType.TRUE_KEYWORD
+import com.pinterest.ktlint.core.ast.ElementType.TYPE_REFERENCE
+import com.pinterest.ktlint.core.ast.ElementType.USER_TYPE
+import com.pinterest.ktlint.core.ast.ElementType.VAL_KEYWORD
+import org.cqfn.diktat.common.config.rules.RulesConfig
+import org.cqfn.diktat.ruleset.constants.Warnings.NULLABLE_PROPERTY_TYPE
+import org.cqfn.diktat.ruleset.utils.KotlinParser
+import org.cqfn.diktat.ruleset.utils.findAllNodesWithSpecificType
+import org.cqfn.diktat.ruleset.utils.hasChildOfType
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.CompositeElement
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LeafPsiElement
+import org.jetbrains.kotlin.com.intellij.psi.tree.IElementType
+
+class NullableTypeRule(private val configRules: List<RulesConfig>) : Rule("nullable-type") {
+
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        isFixMode = autoCorrect
+        emitWarn = emit
+
+        if (node.elementType == PROPERTY)
+            checkProperty(node)
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun checkProperty(node: ASTNode) {
+        if (node.hasChildOfType(VAL_KEYWORD) && node.hasChildOfType(EQ) && node.hasChildOfType(TYPE_REFERENCE)) {
+            if (!node.hasChildOfType(NULL) && node.findChildByType(TYPE_REFERENCE)!!.findAllNodesWithSpecificType(QUEST).isNotEmpty() &&
+                    !node.hasChildOfType(DOT_QUALIFIED_EXPRESSION) && !node.hasChildOfType(CALL_EXPRESSION)) {
```

Comment:
I think that when property is initialized with some collection factory method like `emptyList()` and has nullable type, we still should change type to non-nullable.